### PR TITLE
Fix causatives page

### LINKED ISSUE #3771: Causatives page mixes in matching causatives
At some point it appears we started pulling matching causative onto the CausativeS page. Can be verified locally with e.g. two cases with the same variants.
Screenshot 2023-01-18 at 20 38 26

### PR DESCRIPTION
**Pull request by northwestwitch**
Fix #3771  - Show only causative and not matching causatives

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. Locally, with the 2 RD demo cases.
2. On main branch, go to a variant of a case and mark it as causative
3. Go to causatives page and check that the same variant from the other case is shown (also if it's not causative)
4. Repeat on this branch

**Expected outcome**:
This branch should only show 1 variant on the Causatives page (the original causative)
 
**Review:**
- [x] code approved by DN
- [x] tests executed by DN, CR

### CODE REVIEW

File: scout/adapter/mongo/variant.py
Comment (northwestwitch) [inline scout/adapter/mongo/variant.py:558]: The problem was that when invoked from causatives controller case_obj here is None. Then `match_affected_gt` queries for variants in all cases of the institute_obj, also those that are not causative.

File: tests/server/blueprints/variant/test_variant_controllers.py
Comment (northwestwitch) [inline tests/server/blueprints/variant/test_variant_controllers.py:361]: Was this wrong from the start? 🤔

**Review by dnil — APPROVED** [review]
Tests ok locally, and code looks good for this use case so I approve! If I wasn't at the meeting and this wasn't a bug fix I would try to understand the other entries into the adapter as well. In particular Im a little worried that we are overlooking something when the test outcome changes - though it certainly looks more consistent with the comment now? Perhaps we should also add an issue to investigate/refactor it?

**Comment by northwestwitch** [comment]
> though it certainly looks more consistent with the comment now? Perhaps we should also add an issue to investigate/refactor it?

O think that before merging I'll try to understand why the test was formulated like that. The easiest explanation is that the test was reflecting the pre-existent error in the functionality?

**Comment by northwestwitch** [comment]
> Tests ok locally, and code looks good for this use case so I approve! If I wasn't at the meeting and this wasn't a bug fix I would try to understand the other entries into the adapter as well. In particular Im a little worried that we are overlooking something when the test outcome changes - though it certainly looks more consistent with the comment now? Perhaps we should also add an issue to investigate/refactor it?

Now that I look at the code, looks more likely that the test was correct, and the comment was wrong.

I think it should be like this:
- Case A with a causative 
- Case B with causative in an affected individual
-----> check_causatives returns the causative

- Case A with a causative 
- Case B with causative in an UNaffected individual
-----> check_causatives DOES NOT return the causative

File: scout/adapter/mongo/variant.py
Comment (northwestwitch) [inline scout/adapter/mongo/variant.py:566]: This was an error introduced yesterday by me.

Basically case_obj is the case you are analysing while other_case is the case with the original causative. Of course if you use other case in match_affected_gt it will return the variant, but you are interested in checking of the causative from the other case is found in one affected individual of the case you are analysing... Does it make any sense?

**Comment by northwestwitch** [comment]
Ok, it still doesn't pass the tests. I understood why but perhaps instead of a dirty fix I'll try refactoring to make the code more simple. We have at least 3 functions that return causatives and it's not clear what does what and why. I'll try!

File: tests/server/blueprints/variant/test_variant_controllers.py
Comment (northwestwitch) [reply tests/server/blueprints/variant/test_variant_controllers.py:361]: Nope, the test was right, it was the comment to be wrong!

File: scout/adapter/mongo/variant.py
Comment (northwestwitch) [inline scout/adapter/mongo/variant.py:505]: This is a completely new function called from institute controllers to get variants for the Cauaatives page

File: scout/adapter/mongo/variant.py
Comment (northwestwitch) [inline scout/adapter/mongo/variant.py:541]: This is the old `check_causatives`, that now is used only to retrieve matching causatives from other cases

**Comment by dnil** [comment]
Seems like a good idea to split it. I'll have a look if you wish, but the specified test looks good again ofcourse: 

Before:
Screenshot 2023-01-31 at 13 46 36
After:
Screenshot 2023-01-31 at 13 47 13

**Comment by northwestwitch** [comment]
> Seems like a good idea to split it. I'll have a look if you wish, but the specified test looks good again ofcourse:
> 
> Before: Screenshot 2023-01-31 at 13 46 36 After: Screenshot 2023-01-31 at 13 47 13

Code-wise it is more safe now. I've tested it as well and looked fine. Shall I merge?

**Comment by dnil** [comment]
I've just opened it, and can have a look since you asked before. But yeah, I would definitely think splitting this into two functions instead of having a single multipurpose adapter with many conditionals is safer. The only reason to not duplicate in that case is basically to have better maintainability, but if we dont remember all the uses it has, it can't be maintained anyway!

I guess this means the we've had this for about 5 months then? Or have we had something else hide it? I guess it was just in a period when no-one was wanting causatives stats, or only partial such and didn't notice?

File: scout/adapter/mongo/variant.py
Comment (dnil) [inline scout/adapter/mongo/variant.py:529]: 🥇

File: tests/server/blueprints/variant/test_variant_controllers.py
Comment (dnil) [inline tests/server/blueprints/variant/test_variant_controllers.py:380]: 🥇

File: tests/server/blueprints/variant/test_variant_controllers.py
Comment (dnil) [inline tests/server/blueprints/variant/test_variant_controllers.py:351]: 👀

File: tests/server/blueprints/variant/test_variant_controllers.py
Comment (dnil) [inline tests/server/blueprints/variant/test_variant_controllers.py:355]: 👍

File: scout/adapter/mongo/variant.py
Comment (dnil) [reply scout/adapter/mongo/variant.py:541]: You should probably ditch the `None` defaults here? It would just crash with `case_obj == None` wouldn't it?

**Review by dnil — APPROVED** [review]
Looks better than before, and great with the test clarification. Just a small thing regarding the `case_obj` passing that now looks mandatory to me in the "old" adapter function.

**Comment by dnil** [comment]
> I think you are right, it must have been like this for a few months. At least since when I refactored so that Causatives and Verified share the same code. Perhaps even longer 😟

The salient point I guess being that the risk was/is incidental findings in other cases, but since we haven't heard of any, I suppose we were lucky. 😅 Silly with that `>0` test as well; if that had just been `==1` from the get-go we would probably have caught it.

File: scout/adapter/mongo/variant.py
Comment (northwestwitch) [reply scout/adapter/mongo/variant.py:541]: Right! Thanks!

**Comment by northwestwitch** [comment]
> had just been `==1` from the get-go we would probably have caught it

Yes! 😞

**Pull request merged**
